File: python/problems/leetcode/0300_longest_increasing_subsequence/solution.py
"""LeetCode 300: Longest Increasing Subsequence."""

from __future__ import annotations

from bisect import bisect_left


class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        tails: list[int] = []
        for x in nums:
            i = bisect_left(tails, x)
            if i == len(tails):
                tails.append(x)
            else:
                tails[i] = x
        return len(tails)


if __name__ == "__main__":
    assert Solution().lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18]) == 4
    assert Solution().lengthOfLIS([0, 1, 0, 3, 2, 3]) == 4
    assert Solution().lengthOfLIS([7, 7, 7, 7, 7, 7, 7]) == 1
    print("0300 OK")

File: cpp/problems/leetcode/0300_longest_increasing_subsequence/solution.cpp
// LeetCode 300
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> tails;
        for (int x : nums) {
            auto it = lower_bound(tails.begin(), tails.end(), x);
            if (it == tails.end())
                tails.push_back(x);
            else
                *it = x;
        }
        return static_cast<int>(tails.size());
    }
};

int main() {
    vector<int> a{10, 9, 2, 5, 3, 7, 101, 18};
    assert(Solution().lengthOfLIS(a) == 4);
    vector<int> b{0, 1, 0, 3, 2, 3};
    assert(Solution().lengthOfLIS(b) == 4);
    vector<int> c{7, 7, 7, 7, 7, 7, 7};
    assert(Solution().lengthOfLIS(c) == 1);
    cout << "0300 OK" << endl;
    return 0;
}
